### PR TITLE
Ticket 4165: Load script generator config from file

### PR DESCRIPTION
**Pull request by DominicOram**
### Description of work

Loads information about the configuration of the script generator from a user defined python file.

### Ticket

See https://github.com/ISISComputingGroup/IBEX/issues/4165

### Acceptance criteria

To test:
* Comment out the hidden perspective
* Open the script generator perspective and confirm you can switch configs and the table changes
* Modify the parameters in the python code, restart the GUI and confirm the table updates appropriately.

### Unit tests
Little logic that could unit tested.

### System tests
Should probably do squish tests but ran out of time.

### Documentation
Ran out of time but documentation should be in https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-IBEX-Script-Generator

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by JamesKingWork** [comment]
A note for whoever is reworking and reviewing code on this pull request.

A good amount of the code and config changes that have occurred are due to being merged from https://github.com/ISISComputingGroup/ibex_gui/pull/978 

This code and changed may seem irrelevant in the context of this ticket, but will likely not be in the context of ticket #4166

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        arguments = inspect.getfullargspec(self.action.run).annotations
+        return ListConverter().convert(arguments.keys(), gateway._gateway_client)
+
+    def doAction(self, list_of_arguments):
+        self.action.run(*list_of_arguments)
+
+    def parametersValid(self, list_of_arguments):
+        return self.action.parameters_valid(*list_of_arguments)
+
+    def equals(self, other_config):
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: We are implementing an equals method should we also implement a hashCode method?

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+from py4j.clientserver import ClientServer, JavaParameters, PythonParameters
+from py4j.java_collections import ListConverter
+import inspect
+import argparse
+from action_interface import Action
+import os
+import sys
+
+
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: Methods should have return types even if they are None or Any

Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: There are multiple times where dict is used as a type hint. This should not be done, we should import Dict from the typing module and use that.

Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: We should also attempt to provide the best possible type hinting with Dict by specifying key and value types if possible e.g. Dict[str, TypeVar[T]] or Mapping[str, TypeVar[T]]

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+class Config(object):
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: I'm not certain that the name Config and ConfigWrapper are completely clear as to their purpose. My understanding of a Config was that it contains multiple actions (which the ConfigWrapper actually does). Maybe the naming should be changed to reflect that, unless there is something I have misunderstood.

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ActionParameter.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package uk.ac.stfc.isis.ibex.scriptgenerator;
+
+/**
+ * Data type for an action parameter.
+ * 
+ */
+public class ActionParameter {
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ActionParameter.java]: Possibly, but default values will be implemented in this ticket: https://github.com/ISISComputingGroup/IBEX/issues/4907

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_interface.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from abc import abstractmethod
+
+
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_interface.py]: No, this class defines what an action is and what its parameters are. The `ScriptGeneratorAction` class holds the actual parameter values which will be used to populate this `Action` in part of the script.

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/imat.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from action_interface import Action
+
+class DoRun(Action):
+    def run(self, the: float, imat: float, fields: float, there: float, are: int, more: str):
+        pass
+
+    def parameters_valid(self, the: float, imat: float, fields: float, there: float, are: int, more: str):
+        return ""
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/imat.py]: We discussed this, we should return `None` in the case of no errors.

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/larmor.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+        g.begin()
+        g.waitfor_uamps(uamps)
+        g.end()
+
+    def parameters_valid(self, temperature: float, field: float, uamps: float):
+        if not 0.1 <= temperature <= 300:
+            return "Temperature outside range"
+        if not -5 <= field < 5:
+            return "Field outside range"
+        return ""
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/larmor.py]: Removed

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	 */
+	public ConfigLoader getConfigLoader() {
+		return configLoader;
+	}
+
+	/**
+	 *  Adds the new action parameters to the empty ActionsTable.
+	 * @param actionParameters
+	 */
+	public void setActionParameters(ArrayList<ActionParameter> actionParameters) {
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/ScriptGeneratorSingleton.java]: Yes, this is now private

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/imat.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from action_interface import Action
+
+class DoRun(Action):
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/imat.py]: As I understand it, `imat.py` and `larmor.py` are just example scripts for testing. The instrument scientists are expected to write their own scripts, but would probably not want/care about type hinting.

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/ConfigLoader.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+		return builder.start();
+	}
+	
+	/**
+	 * Constructor for the config loader, initialises the connection with python and reads the configurations.
+	 */
+	public ConfigLoader() {
+        try {
+        	clientServer = createClientServer();
+        	pythonProcess = startPythonProcess(clientServer, "C:\\Instrument\\Apps\\Python3\\python.exe", "/defined_actions/action_loader.py");
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/ConfigLoader.java]: I've added a python 3 preference to the GUI which this now references

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/larmor.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+from action_interface import Action
+
+
+class DoRun(Action):
+    def run(self, temperature: float, field: float, uamps: float):
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/instruments/larmor.py]: These are sample scripts, the real ones will be written by instrument scientists

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	
+	/**
+	 * Container for the UI objects.
+	 * 
+	 * @param parent the parent composite.
+	 */
+	@PostConstruct
+	public void createPartControl(Composite parent) {
+		scriptGeneratorModel = Activator.getModel();
+		configLoader = scriptGeneratorModel.getConfigLoader();
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: A lot of this code is boilerplate from the eclipse framework to draw all the UI elements so is hard to clear up. I have added some breadcrumbs to point out what each section and moved a method into its own function.

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py (hunk 3 above)
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/defined_actions/action_loader.py]: I've added some documentation for this code here: https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-IBEX-Script-Generator#importing-configs-from-user-supplied-actiondefinitions

**Comment by Alistair-McGann-Tessella** [comment]
> 
> 
> We should have squish tests for this.

I've added units tests where possible, but we should write squish tests when the code is somewhat stable.
> 
> Should be documented in https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-IBEX-Script-Generator

👍 
> 
> When there are no defined actions a blank window is produced. Is this the desired behaviour or should we show a message to the user saying no actions have been defined.

I think this is okay, it's the same behaviour as existing script generators and Excel do.
> 
> There are a few things which I have noted in comments that may require a little bit of thought, please do discuss them with me if necessary or if they are unclear etc.

👍

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/PythonInterface.java [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	/**
+	 * Forwards errors from the python process.
+	 */
+	private Runnable listenToErrors = () -> {
+        try {        
+        	InputStreamReader isr = new InputStreamReader(pythonProcess.getErrorStream());
+	        BufferedReader br = new BufferedReader(isr);
+	        String line = null;
+
+			while ((line = br.readLine()) != null) {
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/PythonInterface.java]: This is not as tight a loop as I first thought, as readLine() blocks the thread until an error comes in through the stream. John demonstrated this by putting breakpoints on this line and the line below it, which are only triggered once in normal operation (no python errors)

[the remark this answers is not in the document]
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/pythoninterface/PythonInterface.java]: That is fine then, I hadn't realised that was the behaviour

File: base/uk.ac.stfc.isis.ibex.client.tycho.parent/pom.xml [hunk 14]
```diff
@@ -240,6 +239,11 @@
 		<module>../uk.ac.stfc.isis.ibex.alarm.tests</module>
 		<module>../uk.ac.stfc.isis.ibex.ui.scripting.tests</module>
 		-->
+		
+		<!-- These tests currently fail in surefire because the python interface is loaded on class instantiation
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.client.tycho.parent/pom.xml]: Done in #4759

**Pull request merged**